pkg/sys/unshare_unsafe_linux.go

```diff
@@ -0,0 +1,187 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sys
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"runtime"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const reaperTimeout = 60 * time.Second
+
+// UnshareAfterEnterUserns allows to disassociate parts of its execution context
+// within a user namespace.
+func UnshareAfterEnterUserns(uidMap, gidMap string, unshareFlags uintptr, f func(pid int) error) error {
+	if unshareFlags&syscall.CLONE_NEWUSER == syscall.CLONE_NEWUSER {
+		return fmt.Errorf("unshare flags should not include user namespace")
+	}
+
+	usernsFd, err := GetUsernsFD(uidMap, gidMap)
+	if err != nil {
+		return fmt.Errorf("failed to create user namespace: %w", err)
+	}
+	defer usernsFd.Close()
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	pr, pw, err := os.Pipe()
+	if err != nil {
+		return fmt.Errorf("failed to open pipe: %w", err)
+	}
+	defer func() {
+		pr.Close()
+		pw.Close()
+	}()
+
+	pid, pidfd, errno := unshareAfterEnterUserns(usernsFd.Fd(), unshareFlags, pw.Fd())
+	if errno != 0 {
+		return errno
+	}
+
+	exitCh := make(chan struct{})
+
+	pidFd := os.NewFile(pidfd, "pidfd")
+	defer pidFd.Close()
+
+	defer func() {
+		unix.PidfdSendSignal(int(pidFd.Fd()), unix.SIGKILL, nil, 0)
+
+		ctx, cancel := context.WithTimeout(context.Background(), reaperTimeout)
+		defer cancel()
+
+		select {
+		case <-exitCh:
+			// try to drain extra goroutine
+		case <-ctx.Done():
+			panic(fmt.Errorf("failed to reap zombie in %s", reaperTimeout))
+		}
+	}()
+
+	go func() {
+		defer close(exitCh)
+		// Close pipe if the child process exists. So, the read on pipe
+		// will receive EOF event.
+		defer pw.Close()
+
+		pidfdWaitid(pidFd)
+	}()
+
+	buf := make([]byte, 1)
+	_, err = pr.Read(buf)
+	if err != nil {
+		return fmt.Errorf("failed to receive ready notification: %w", err)
+	}
+	if buf[0] != byte(1) {
+		return fmt.Errorf("expected ready notification 1, but got %v", buf[0])
+	}
+
+	if f == nil {
+		return nil
+	}
+	return f(int(pid))
+}
+
+// unshareAfterEnterUserns clones child process first, associates that process
+// with existing user namespace and then unshares into new given namespaces.
+//
+// Precondition: The runtime OS thread must be locked, which is GO runtime
+// requirement.
+//
+//go:norace
+//go:noinline
+//go:nosplit
+func unshareAfterEnterUserns(usernsFd uintptr, flags uintptr, pipeFd uintptr) (_pid uintptr, _pidfd uintptr, _ syscall.Errno) {
```

Comment:
Why did you change it to this? The one creating a process seem very easy to reason about

Reply:
Do you mean using `os.StartProcess`? Sorry I force-push to clean-up the history.

Yes. I have plan to use `os.StartProcess(ptrace: true)`. But I run into two problems:

1. https://github.com/containerd/containerd/pull/10611#issuecomment-2296743964
2. 
```
2024-08-19T13:29:27.6655085Z     pod_userns_linux_test.go:284: 
2024-08-19T13:29:27.6656082Z         	Error Trace:	/home/runner/work/containerd/containerd/integration/pod_userns_linux_test.go:284
2024-08-19T13:29:27.6656870Z         	Error:      	Received unexpected error:
2024-08-19T13:29:27.6662188Z         	            	rpc error: code = Unknown desc = failed to update container "16dd29b770c6a7127b84ebf1228617b417498522281b391ef5bd35d05ea6c85e" state: failed to checkpoint status to "/var/lib/containerd-test/io.containerd.grpc.v1.cri/containers/16dd29b770c6a7127b84ebf1228617b417498522281b391ef5bd35d05ea6c85e/status": close /var/lib/containerd-test/io.containerd.grpc.v1.cri/containers/16dd29b770c6a7127b84ebf1228617b417498522281b391ef5bd35d05ea6c85e/.tmp-status359157146: bad file descriptor
2024-08-19T13:29:27.6666867Z         	Test:       	TestPodUserNS/volumes_permissions
```

I can't reproduce it in my local. Maybe related to go1.23.0. It seems that the go runtime randomly closes fd.

Reply:
Will switch to use `os.StartProcess` if there is any workrounds from https://github.com/golang/go/issues/68984, like dup or reset finalizer.

Reply:
Oh, great finding! Wouldn't just doing a dup here be a valid workaround using `unix.Dup()`?

Reply:
Will update it after https://github.com/containerd/containerd/pull/10611 merged